src/Monitoring/Sdk/GrafanaSerialization.cs

```diff
@@ -108,6 +118,118 @@ public static JObject SanitizeNotificationChannel(JObject notificationChannel)
+
+            // These paths may be parameterized
+            string[] paths = new[]
+            {
+                "$.dashboard.panels[*].targets[*].azureMonitor.resourceGroup",
+                "$.dashboard.panels[*].targets[*].azureMonitor.resourceName",
+                "$.dashboard.panels[*].targets[*].azureLogAnalytics.resource",
+                "$.dashboard.panels[*].targets[*].azureLogAnalytics.workspace"
+            };
+
+            foreach (string path in paths)
+            {
+                IEnumerable<JToken> tokens = parameterizedDashboard.SelectTokens(path);
+
+                foreach (JToken token in tokens)
+                {
+                    string value = token.Value<string>();
+
+                    // Value Can technically be other JSON types (but the paths specified above
+                    // will always be strings). Because the object must be a "string" to store the
+                    // parameter info, supporting anything other than a string will require also
+                    // keeping type info. So only support strings for now.
+
+                    if (value == null)
+                    {
+                        throw new Exception("Unexpected token type. Value must be a string.");
+                    }
+
+                    Parameter p = parameters
+                        .Where(p => p.Values.ContainsKey(activeEnvironment) && p.Values[activeEnvironment] == value)
+                        .FirstOrDefault();
+
+                    if (p == null)
+                    {
+                        string name = Guid.NewGuid().ToString();
+
+                        p = new Parameter()
+                        {
+                            Name = name,
+                            Values = environments.Select(env =>
+                            {
+                                if (env == activeEnvironment)
+                                {
+                                    return (env, value);
+                                }
+                                else
+                                {
+                                    return (env, "PLACEHOLDER");
+                                }
+                            }).ToDictionary(e => e.env, e => e.Item2)
+                        };
+
+                        parameters.Add(p);
+                    }
+
+                    token.Replace(new JValue($"[parameter({p.Name})]"));
+                }
+            }
+
+            return parameterizedDashboard;
+        }
+
+        public static JObject DeparameterizeDashboard(JObject dashboard, IEnumerable<Parameter> parameters, string environment)
+        {
+            JObject deparameterizedDashboard = new JObject(dashboard);
+
+            IEnumerable<JToken> tokens = deparameterizedDashboard.SelectTokens("$..*")
+                .Where(token => token.Type == JTokenType.String)
+                .Where(token => TryGetParameterName(token.Value<string>(), out string _))
```

Comment:
Why not just have this inside the loop? That way you don't need to parse the token twice.  Loop over every token (since that's what this "where" is already doing.  If you can't get a parameter out, then continue, otherwise do the replacement.

Reply:
Good eye.